cmd/syncthing/upnpsvc.go

```diff
@@ -0,0 +1,108 @@
+// Copyright (C) 2015 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package main
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/syncthing/syncthing/internal/config"
+	"github.com/syncthing/syncthing/internal/upnp"
+)
+
+// The UPnP service runs a loop for discovery of IGDs (Internet Gateway
+// Devices) and setup/renewal of a port mapping.
+type upnpSvc struct {
+	cfg       *config.Wrapper
+	localPort int
+}
+
+func newUPnPSvc(cfg *config.Wrapper, localPort int) *upnpSvc {
+	return &upnpSvc{
+		cfg:       cfg,
+		localPort: localPort,
+	}
+}
+
+func (s *upnpSvc) Serve() {
+	extPort := 0
+	noIGDFound := false
+
+	for {
+		igds := upnp.Discover(time.Duration(s.cfg.Options().UPnPTimeoutS) * time.Second)
+		if len(igds) >= 0 {
+			noIGDFound = false
+			extPort = s.tryIGDs(igds, extPort)
+		} else if !noIGDFound {
+			noIGDFound = true
+			l.Infof("No UPnP device detected")
```

Comment:
No, only first time, or first time after we had previously found something. :)

(Or that's the intention of the crappily named `noIGDFound`)